odl/space/fspace.py

```diff
@@ -702,17 +702,24 @@ def astype(self, out_dtype):
         if out_dtype == self.out_dtype:
             return self
 
-        # Caching for real and complex versions (exact dtyoe mappings)
-        if out_dtype == self.real_out_dtype:
-            if self.__real_space is None:
-                self.__real_space = self._astype(out_dtype)
-            return self.__real_space
-        elif out_dtype == self.complex_out_dtype:
-            if self.__complex_space is None:
-                self.__complex_space = self._astype(out_dtype)
-            return self.__complex_space
-        else:
+        # Try to use caching for real and complex versions (exact dtype
+        # mappings). This may fail for certain dtype, in which case we
+        # just go to `_astype` directly.
+        try:
+            real_dtype = self.real_out_dtype
+        except TypeError:
```

Comment:
Why typeerror here? Shouldn't this be an attributeerror?

Reply:
I'm reluctant to use `AttributeError` for this kind of stuff. To me it feels like `attr = getattr(obj, 'attr', None)` should be equivalent to
```python
try:
    attr = obj.attr
except AttributeError:
    attr = None
```
which it wouldn't if I raised an `AttributeError` for an existing attribute.